File: java/lib/src/main/java/algorithm/sorting/QuickSort.java
package algorithm.sorting;

public class QuickSort extends Sort {
  private int partition(int[] arr, int low, int high, SortType type) {
    int pivot = arr[high];

    int i = low - 1, j = high;

    while (true) {
      while (super.compare(pivot, arr[++i], type))
        ;

      while (i < --j && super.compare(arr[j], pivot, type))
        ;

      if (j <= i) {
        break;
      }
      super.swap(arr, i, j);
    }

    super.swap(arr, i, high);
    return i;
  }

  private int[] quickSort(int[] arr, int low, int high, SortType type) {
    if (low < high) {
      int pivot = partition(arr, low, high, type);
      quickSort(arr, low, pivot - 1, type);
      quickSort(arr, pivot + 1, high, type);
    }

    return arr;
  }

  @Override
  protected int[] sort(int[] arr, SortType type) {
    int length = arr.length;
    return quickSort(arr, 0, length - 1, type);
  }
}

File: kotlin/lib/src/main/kotlin/algorithm/sorting/QuickSort.kt
package algorithm.sorting

public class QuickSort : Sort() {
  override fun sort(arr: Array<Int>, type: SortType): Array<Int> {
    val length = arr.size
    return quickSort(arr, 0, length - 1, type)
  }

  fun partition(arr: Array<Int>, low: Int, high: Int, type: SortType): Int {
    val pivot = arr[high]
    var i = low - 1
    var j = high
    while (true) {
      while (compare(pivot, arr[++i], type)) ;
      while (i < --j && compare(arr[j], pivot, type)) ;
      if (i >= j) break
      arr.swap(i, j)
    }
    arr.swap(i, high)
    return i
  }

  fun quickSort(arr: Array<Int>, low: Int, high: Int, type: SortType): Array<Int> {
    if (low < high) {
      val pivot = partition(arr, low, high, type)
      quickSort(arr, low, pivot - 1, type)
      quickSort(arr, pivot + 1, high, type)
    }

    return arr
  }
}
